src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size() || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+        if (index == 0 && size() == 0) {
+            add(value);
+        } else if (index == size()) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = searchNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+        Node<T> getNode = head;
+        for (int i = 0; i < index; i++) {
+            getNode = getNode.next;
+            if (getNode == null) {
+                throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+            }
+        }
+        return getNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T removeValue;
+        if (index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+            if (currentNode == null) {
+                throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+            }
+        }
+        removeValue = currentNode.value;
+        currentNode.value = value;
+        return removeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+        Node<T> currentNode = searchNode(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size();i++) {
+            if (object == null || currentNode.value == null) {
+                if (object == currentNode.value) {
+                    unlink(currentNode);
+                    return true;
+                }
+            } else if (object.equals(currentNode.value)) {
+                unlink(currentNode);
+                return true;
+            }
```

Comment:
you have the same body in both cases
just combine ifs on lines 97, 102 - look up code for Objects.equals, but remember that you can't use it directly